### PR TITLE
Port #424 from yuzu: "string_util: Remove StringFromFormat() and related functions"

### PR DESCRIPTION
**Pull request by FearlessTobi**
See yuzu-emu/yuzu#424 for more details. 

**NOTE**: I'm 100% sure I've done something wrong in at least one place in this PR, so please review closely.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4147)

### CODE REVIEW

File: src/core/file_sys/archive_extsavedata.cpp [hunk 1]
```diff
@@ -165,16 +166,15 @@ std::string GetExtSaveDataPath(const std::string& mount_point, const Path& path)
     ExtSaveDataArchivePath path_data;
     std::memcpy(&path_data, vec_data.data(), sizeof(path_data));
 
-    return Common::StringFromFormat("%s%08X/%08X/", mount_point.c_str(), path_data.save_high,
-                                    path_data.save_low);
+    return fmt::format("{}{:08x}/{:08x}/", mount_point.c_str(), path_data.save_high,
```
Comment (MerryMage) [inline src/core/file_sys/archive_extsavedata.cpp]: `c_str()` unnecessary (ditto below)

File: src/core/file_sys/archive_source_sd_savedata.cpp [hunk 2]
```diff
@@ -18,22 +19,19 @@ namespace FileSys {
 namespace {
 
 std::string GetSaveDataContainerPath(const std::string& sdmc_directory) {
-    return Common::StringFromFormat("%sNintendo 3DS/%s/%s/title/", sdmc_directory.c_str(),
-                                    SYSTEM_ID, SDCARD_ID);
+    return fmt::format("{}Nintendo 3DS/{}/{}/title/", sdmc_directory.c_str(), SYSTEM_ID, SDCARD_ID);
```
Comment (MerryMage) [inline src/core/file_sys/archive_source_sd_savedata.cpp]: `c_str()` unnecessary (ditto below)

File: src/core/file_sys/archive_systemsavedata.cpp [hunk 3]
```diff
@@ -25,11 +26,11 @@ std::string GetSystemSaveDataPath(const std::string& mount_point, const Path& pa
     u32 save_high;
     std::memcpy(&save_low, &vec_data[4], sizeof(u32));
     std::memcpy(&save_high, &vec_data[0], sizeof(u32));
-    return Common::StringFromFormat("%s%08X/%08X/", mount_point.c_str(), save_low, save_high);
+    return fmt::format("{}{:08X}/{:08X}/", mount_point.c_str(), save_low, save_high);
```
Comment (MerryMage) [inline src/core/file_sys/archive_systemsavedata.cpp]: `c_str()` unnecessary (ditto below)

File: src/core/hle/service/am/am.cpp [hunk 4]
```diff
@@ -417,16 +418,15 @@ std::string GetTitleContentPath(Service::FS::MediaType media_type, u64 tid, u16
         }
     }
 
-    return Common::StringFromFormat("%s%08x.app", content_path.c_str(), content_id);
+    return fmt::format("{}{:08x}.app", content_path.c_str(), content_id);
```
Comment (MerryMage) [inline src/core/hle/service/am/am.cpp]: `c_str()` unnecessary (ditto below)

File: src/core/hle/kernel/svc.cpp [hunk 5]
```diff
@@ -723,7 +724,7 @@ static ResultCode GetResourceLimitLimitValues(VAddr values, Handle resource_limi
 /// Creates a new thread
 static ResultCode CreateThread(Handle* out_handle, u32 priority, u32 entry_point, u32 arg,
                                u32 stack_top, s32 processor_id) {
-    std::string name = Common::StringFromFormat("unknown-%08" PRIX32, entry_point);
+    std::string name = fmt::format("unknown-{:08X}" PRIX32, entry_point);
```
Comment (wwylele) [inline src/core/hle/kernel/svc.cpp]: Remove PRIX32

File: src/core/file_sys/archive_extsavedata.cpp [hunk 6]
```diff
@@ -165,16 +166,14 @@ std::string GetExtSaveDataPath(const std::string& mount_point, const Path& path)
     ExtSaveDataArchivePath path_data;
     std::memcpy(&path_data, vec_data.data(), sizeof(path_data));
 
-    return Common::StringFromFormat("%s%08X/%08X/", mount_point.c_str(), path_data.save_high,
-                                    path_data.save_low);
+    return fmt::format("{}{:08x}/{:08x}/", mount_point, path_data.save_high, path_data.save_low);
```
Comment (MerryMage) [inline src/core/file_sys/archive_extsavedata.cpp]: Capital `X`. Changing the case may break people on case-sensitive file systems.

Comment (FearlessTobi) [reply src/core/file_sys/archive_extsavedata.cpp]: Fixed.

File: src/citra/emu_window/emu_window_sdl2.cpp [hunk 7]
```diff
@@ -100,8 +101,8 @@ EmuWindow_SDL2::EmuWindow_SDL2(bool fullscreen) {
     SDL_GL_SetAttribute(SDL_GL_BLUE_SIZE, 8);
     SDL_GL_SetAttribute(SDL_GL_ALPHA_SIZE, 0);
 
-    std::string window_title = Common::StringFromFormat(
-        "Citra %s| %s-%s ", Common::g_build_fullname, Common::g_scm_branch, Common::g_scm_desc);
+    std::string window_title = fmt::format("Citra {} | {}-{}", Common::g_build_fullname,
```
Comment (zhaowenlan1779) [inline src/citra/emu_window/emu_window_sdl2.cpp]: extra space after the first `{}`

File: src/citra_qt/bootmanager.cpp [hunk 8]
```diff
@@ -103,8 +105,8 @@ class GGLWidgetInternal : public QGLWidget {
 GRenderWindow::GRenderWindow(QWidget* parent, EmuThread* emu_thread)
     : QWidget(parent), child(nullptr), emu_thread(emu_thread) {
 
-    std::string window_title = Common::StringFromFormat("Citra %s| %s-%s", Common::g_build_name,
-                                                        Common::g_scm_branch, Common::g_scm_desc);
+    std::string window_title = fmt::format("Citra {} | {}-{}", Common::g_build_name,
```
Comment (zhaowenlan1779) [inline src/citra_qt/bootmanager.cpp]: ditto

File: src/common/timer.cpp [hunk 9]
```diff
@@ -3,6 +3,9 @@
 // Refer to the license.txt file included.
 
 #include <ctime>
+
+#include <fmt/format.h>
+
```
Comment (zhaowenlan1779) [inline src/common/timer.cpp]: we do not split `#include`s

File: src/core/hle/kernel/svc.cpp [hunk 10]
```diff
@@ -825,7 +826,7 @@ static ResultCode SetThreadPriority(Handle handle, u32 priority) {
 /// Create a mutex
 static ResultCode CreateMutex(Handle* out_handle, u32 initial_locked) {
     SharedPtr<Mutex> mutex = Mutex::Create(initial_locked != 0);
-    mutex->name = Common::StringFromFormat("mutex-%08x", Core::CPU().GetReg(14));
+    mutex->name = fmt::format("mutex-{:08X}", Core::CPU().GetReg(14));
```
Comment (zhaowenlan1779) [inline src/core/hle/kernel/svc.cpp]: should this be `x` (lower case) instead? (do not really matter, but you may want to keep the same behavior)

File: src/core/hle/kernel/svc.cpp [hunk 11]
```diff
@@ -888,7 +889,7 @@ static ResultCode GetThreadId(u32* thread_id, Handle handle) {
 /// Creates a semaphore
 static ResultCode CreateSemaphore(Handle* out_handle, s32 initial_count, s32 max_count) {
     CASCADE_RESULT(SharedPtr<Semaphore> semaphore, Semaphore::Create(initial_count, max_count));
-    semaphore->name = Common::StringFromFormat("semaphore-%08x", Core::CPU().GetReg(14));
+    semaphore->name = fmt::format("semaphore-{:08X}", Core::CPU().GetReg(14));
```
Comment (zhaowenlan1779) [inline src/core/hle/kernel/svc.cpp]: ditto

File: src/core/hle/kernel/svc.cpp [hunk 12]
```diff
@@ -938,9 +939,8 @@ static ResultCode QueryMemory(MemoryInfo* memory_info, PageInfo* page_info, u32
 
 /// Create an event
 static ResultCode CreateEvent(Handle* out_handle, u32 reset_type) {
-    SharedPtr<Event> evt =
-        Event::Create(static_cast<ResetType>(reset_type),
-                      Common::StringFromFormat("event-%08x", Core::CPU().GetReg(14)));
+    SharedPtr<Event> evt = Event::Create(static_cast<ResetType>(reset_type),
+                                         fmt::format("event-{:08X}", Core::CPU().GetReg(14)));
```
Comment (zhaowenlan1779) [inline src/core/hle/kernel/svc.cpp]: ditto

File: src/core/hle/kernel/svc.cpp [hunk 13]
```diff
@@ -982,9 +982,8 @@ static ResultCode ClearEvent(Handle handle) {
 
 /// Creates a timer
 static ResultCode CreateTimer(Handle* out_handle, u32 reset_type) {
-    SharedPtr<Timer> timer =
-        Timer::Create(static_cast<ResetType>(reset_type),
-                      Common::StringFromFormat("timer-%08x", Core::CPU().GetReg(14)));
+    SharedPtr<Timer> timer = Timer::Create(static_cast<ResetType>(reset_type),
+                                           fmt ::format("timer-{:08X}", Core::CPU().GetReg(14)));
```
Comment (zhaowenlan1779) [inline src/core/hle/kernel/svc.cpp]: ditto

File: src/core/hle/service/service.cpp [hunk 14]
```diff
@@ -125,10 +125,9 @@ static std::string MakeFunctionString(const char* name, const char* port_name,
     // Number of params == bits 0-5 + bits 6-11
     int num_params = (cmd_buff[0] & 0x3F) + ((cmd_buff[0] >> 6) & 0x3F);
 
-    std::string function_string =
-        Common::StringFromFormat("function '%s': port=%s", name, port_name);
+    std::string function_string = fmt::format("function '{}': port={}", name, port_name);
     for (int i = 1; i <= num_params; ++i) {
-        function_string += Common::StringFromFormat(", cmd_buff[%i]=0x%X", i, cmd_buff[i]);
+        function_string += fmt::format(", cmd_buff[{}]={:#X}", i, cmd_buff[i]);
```
Comment (zhaowenlan1779) [inline src/core/hle/service/service.cpp]: The `common/string_util.h` include in this file can be removed.

File: src/core/hle/kernel/svc.cpp (hunk 10 above)
Comment (zhaowenlan1779) [reply src/core/hle/kernel/svc.cpp]: The `common/string_util.h` include in this file can be removed.

File: src/core/gdbstub/gdbstub.cpp [hunk 15]
```diff
@@ -610,16 +611,17 @@ static void SendSignal(Kernel::Thread* thread, u32 signal, bool full = true) {
...
     if (full) {
-        buffer = Common::StringFromFormat("T%02x%02x:%08x;%02x:%08x;%02x:%08x", latest_signal,
-                                          PC_REGISTER, htonl(Core::CPU().GetPC()), SP_REGISTER,
-                                          htonl(Core::CPU().GetReg(SP_REGISTER)), LR_REGISTER,
-                                          htonl(Core::CPU().GetReg(LR_REGISTER)));
+
+        buffer = fmt::format("T{:02x}{:02x}:{:08x};{:02x}:{:08x};{:02x}:{:08x}", latest_signal,
+                             PC_REGISTER, htonl(Core::CPU().GetPC()), SP_REGISTER,
+                             htonl(Core::CPU().GetReg(SP_REGISTER)), LR_REGISTER,
+                             htonl(Core::CPU().GetReg(LR_REGISTER)));
```
Comment (zhaowenlan1779) [inline src/core/gdbstub/gdbstub.cpp]: The `common/string_util.h` include in this file can be removed.

File: src/core/file_sys/archive_systemsavedata.cpp [hunk 16]
```diff
@@ -25,11 +26,11 @@ std::string GetSystemSaveDataPath(const std::string& mount_point, const Path& pa
     u32 save_high;
     std::memcpy(&save_low, &vec_data[4], sizeof(u32));
     std::memcpy(&save_high, &vec_data[0], sizeof(u32));
-    return Common::StringFromFormat("%s%08X/%08X/", mount_point.c_str(), save_low, save_high);
+    return fmt::format("{}{:08X}/{:08X}/", mount_point, save_low, save_high);
```
Comment (zhaowenlan1779) [inline src/core/file_sys/archive_systemsavedata.cpp]: ditto

File: src/core/file_sys/archive_source_sd_savedata.cpp [hunk 17]
```diff
@@ -18,22 +19,19 @@ namespace FileSys {
 namespace {
 
 std::string GetSaveDataContainerPath(const std::string& sdmc_directory) {
-    return Common::StringFromFormat("%sNintendo 3DS/%s/%s/title/", sdmc_directory.c_str(),
-                                    SYSTEM_ID, SDCARD_ID);
+    return fmt::format("{}Nintendo 3DS/{}/{}/title/", sdmc_directory, SYSTEM_ID, SDCARD_ID);
```
Comment (zhaowenlan1779) [inline src/core/file_sys/archive_source_sd_savedata.cpp]: ditto

File: src/core/file_sys/archive_extsavedata.cpp [hunk 18]
```diff
@@ -165,16 +166,14 @@ std::string GetExtSaveDataPath(const std::string& mount_point, const Path& path)
     ExtSaveDataArchivePath path_data;
     std::memcpy(&path_data, vec_data.data(), sizeof(path_data));
 
-    return Common::StringFromFormat("%s%08X/%08X/", mount_point.c_str(), path_data.save_high,
-                                    path_data.save_low);
+    return fmt::format("{}{:08X}/{:08X}/", mount_point, path_data.save_high, path_data.save_low);
```
Comment (zhaowenlan1779) [inline src/core/file_sys/archive_extsavedata.cpp]: ditto

File: src/common/timer.cpp [hunk 19]
```diff
@@ -3,6 +3,9 @@
...
 #include "common/common_types.h"
 #include "common/string_util.h"
```
Comment (zhaowenlan1779) [inline src/common/timer.cpp]: This `#include` can be removed.

File: src/citra_qt/main.cpp [hunk 20]
```diff
@@ -920,7 +921,7 @@ void GMainWindow::OnGameListNavigateToGamedbEntry(
     auto it = std::find_if(
         compatibility_list.begin(), compatibility_list.end(),
         [program_id](const std::pair<std::string, std::pair<QString, QString>>& element) {
-            std::string pid = Common::StringFromFormat("%016" PRIX64, program_id);
+            std::string pid = fmt::format("{:016X}", program_id);
```
Comment (zhaowenlan1779) [inline src/citra_qt/main.cpp]: The `common/string_util.h` include in this file can be removed.

File: src/citra_qt/game_list.cpp [hunk 21]
```diff
@@ -446,7 +447,7 @@ void GameList::AddGamePopup(QMenu& context_menu, u64 program_id) {
     auto it = std::find_if(
         compatibility_list.begin(), compatibility_list.end(),
         [program_id](const std::pair<std::string, std::pair<QString, QString>>& element) {
-            std::string pid = Common::StringFromFormat("%016" PRIX64, program_id);
+            std::string pid = fmt::format("{:016X}", program_id);
```
Comment (zhaowenlan1779) [inline src/citra_qt/game_list.cpp]: ditto

File: src/citra_qt/bootmanager.cpp (hunk 8 above)
Comment (zhaowenlan1779) [reply src/citra_qt/bootmanager.cpp]: The `common/string_util.h` include in this file can be removed.

File: src/citra/emu_window/emu_window_sdl2.cpp (hunk 7 above)
Comment (zhaowenlan1779) [reply src/citra/emu_window/emu_window_sdl2.cpp]: The `common/string_util.h` include in this file can be removed.

Comment (FearlessTobi) [reply src/citra/emu_window/emu_window_sdl2.cpp]: yuzu has this extra space too...

Comment (zhaowenlan1779) [reply src/citra/emu_window/emu_window_sdl2.cpp]: But it is not good and will leave an extra space in the title bar
you should upstream it back to yuzu...

**Comment by FearlessTobi** [comment]
Conflicts are fixed. PR should be good to merge now.

**Pull request merged**